File: src/main/java/ru/nucodelabs/gem/app/snapshot/Snapshot.java
package ru.nucodelabs.gem.app.snapshot;

public interface Snapshot<T> {

    static <T> Snapshot<T> of(T value) {
        return new SnapshotImpl<>(value);
    }

    T value();

    interface Originator<T> {
        Snapshot<T> getSnapshot();

        void restoreFromSnapshot(Snapshot<T> snapshot);
    }
}

File: src/main/java/ru/nucodelabs/gem/app/snapshot/Snapshot.kt
package ru.nucodelabs.gem.app.snapshot;

import java.util.Objects;

public interface Snapshot<T> {

    static <T> Snapshot<T> of(T value) {
        return new SnapshotImpl<>(
                Objects.requireNonNull(value));
    }

    T value();

    interface Originator<T> {
        Snapshot<T> getSnapshot();

        void restoreFromSnapshot(Snapshot<T> snapshot);
    }
}
